File: qabelbox/src/main/java/de/qabel/qabelbox/dagger/modules/StorageModule.java
package de.qabel.qabelbox.dagger.modules;

import android.content.Context;

import java.io.File;
import java.sql.Connection;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import de.qabel.box.storage.FileMetadataFactory;
import de.qabel.box.storage.jdbc.JdbcFileMetadataFactory;
import de.qabel.chat.repository.ChatShareRepository;
import de.qabel.chat.service.MainSharingService;
import de.qabel.chat.service.SharingService;
import de.qabel.core.crypto.CryptoUtils;
import de.qabel.core.repositories.AndroidVersionAdapter;
import de.qabel.core.repository.ContactRepository;
import de.qabel.core.repository.sqlite.VersionAdapter;
import de.qabel.qabelbox.box.provider.DocumentIdParser;
import de.qabel.qabelbox.config.AppPreference;
import de.qabel.qabelbox.storage.notifications.AndroidStorageNotificationManager;
import de.qabel.qabelbox.storage.notifications.AndroidStorageNotificationPresenter;
import de.qabel.qabelbox.storage.notifications.StorageNotificationManager;
import de.qabel.qabelbox.storage.server.AndroidBlockServer;
import de.qabel.qabelbox.storage.server.BlockServer;
import kotlin.jvm.functions.Function1;

@Module
public class StorageModule {

    @Singleton
    @Provides
    StorageNotificationManager providesStorageNotificationManager(Context context) {
        return new AndroidStorageNotificationManager(new AndroidStorageNotificationPresenter(context));
    }

    @Singleton
    @Provides
    DocumentIdParser providesDocumentIdParser() {
        return new DocumentIdParser();
    }

    @Provides
    File providesCacheDir(Context context) {
        return context.getCacheDir();
    }

    @Singleton
    @Provides
    BlockServer providesBlockServer(AppPreference preference, Context context) {
        return createBlockServer(preference, context);
    }

    protected BlockServer createBlockServer(AppPreference preference, Context context){
        return new AndroidBlockServer(preference, context);
    }

    @Singleton
    @Provides
    FileMetadataFactory provideFileMetadataFactory(File cacheDir) {
        return new JdbcFileMetadataFactory(cacheDir, new Function1<Connection, VersionAdapter>() {
            @Override
            public VersionAdapter invoke(Connection connection) {
                return new AndroidVersionAdapter(connection);
            }
        });
    }

    @Singleton
    @Provides
    SharingService providesSharingService(ChatShareRepository shareRepository,
                                          ContactRepository contactRepository,
                                          File cacheDir,
                                          FileMetadataFactory fileMetadataFactory) {
        return new MainSharingService(
                shareRepository,
                contactRepository,
                cacheDir,
                fileMetadataFactory,
                new CryptoUtils());
    }

}

File: qabelbox/src/main/java/de/qabel/qabelbox/dagger/modules/StorageModule.kt
package de.qabel.qabelbox.dagger.modules

import android.content.Context
import dagger.Module
import dagger.Provides
import de.qabel.box.storage.FileMetadataFactory
import de.qabel.box.storage.jdbc.JdbcFileMetadataFactory
import de.qabel.chat.repository.ChatShareRepository
import de.qabel.chat.service.MainSharingService
import de.qabel.chat.service.SharingService
import de.qabel.client.box.BoxSchedulers
import de.qabel.client.box.MainBoxSchedulers
import de.qabel.client.box.documentId.DocumentIdParser
import de.qabel.client.box.interactor.BoxVolumeManager
import de.qabel.client.box.interactor.VolumeManager
import de.qabel.client.box.storage.LocalStorage
import de.qabel.core.crypto.CryptoUtils
import de.qabel.core.repositories.AndroidVersionAdapter
import de.qabel.core.repository.ContactRepository
import de.qabel.core.repository.IdentityRepository
import de.qabel.qabelbox.box.BoxScheduler
import de.qabel.qabelbox.box.interactor.*
import de.qabel.qabelbox.box.notifications.AndroidStorageNotificationManager
import de.qabel.qabelbox.box.notifications.AndroidStorageNotificationPresenter
import de.qabel.qabelbox.box.notifications.StorageNotificationManager
import de.qabel.qabelbox.config.AppPreference
import de.qabel.qabelbox.storage.server.AndroidBlockServer
import de.qabel.qabelbox.storage.server.BlockServer
import rx.schedulers.Schedulers
import java.io.File
import java.util.concurrent.Executors
import javax.inject.Singleton

@Module
open class StorageModule {

    @Singleton
    @Provides
    internal fun providesStorageNotificationManager(context: Context): StorageNotificationManager {
        return AndroidStorageNotificationManager(AndroidStorageNotificationPresenter(context))
    }

    @Singleton
    @Provides
    internal fun providesDocumentIdParser(): DocumentIdParser {
        return DocumentIdParser()
    }

    @Provides
    internal fun providesCacheDir(context: Context): File {
        return context.cacheDir
    }

    @Singleton
    @Provides
    internal fun providesBlockServer(preference: AppPreference, context: Context): BlockServer {
        return createBlockServer(preference, context)
    }

    protected open fun createBlockServer(preference: AppPreference, context: Context): BlockServer {
        return AndroidBlockServer(preference, context)
    }

    @Singleton
    @Provides
    internal fun provideFileMetadataFactory(cacheDir: File): FileMetadataFactory {
        return JdbcFileMetadataFactory(cacheDir, ::AndroidVersionAdapter, JdbcPrefix.jdbcPrefix)
    }

    @Singleton
    @Provides
    internal fun providesSharingService(shareRepository: ChatShareRepository,
                                        contactRepository: ContactRepository,
                                        cacheDir: File,
                                        fileMetadataFactory: FileMetadataFactory): SharingService {
        return MainSharingService(
                shareRepository,
                contactRepository,
                cacheDir,
                fileMetadataFactory,
                CryptoUtils())
    }

    @Provides
    @Singleton
    internal fun providesScheduler(): BoxSchedulers {
        return MainBoxSchedulers(Schedulers.from(Executors.newCachedThreadPool()))
    }

    @Singleton
    @Provides
    fun providesDocumentIdUseCase(useCase: BoxDocumentIdInteractor): DocumentIdInteractor {
        return useCase
    }

    @Singleton
    @Provides
    fun provideVolumeManager(identityRepository: IdentityRepository,
                             contactRepository: ContactRepository,
                             preference: AppPreference,
                             context: Context, blockServer: BlockServer,
                             scheduler: BoxSchedulers,
                             localStorage: LocalStorage):
            VolumeManager {
        val (read, operation) = makeFileBrowserFactory(
                identityRepository, contactRepository, preference.deviceId,
                context.cacheDir, blockServer, scheduler, localStorage)
        return BoxVolumeManager(identityRepository, read, operation)
    }

    @Singleton
    @Provides
    fun providesBoxServiceStarter(serviceStarter: AndroidBoxServiceStarter): BoxServiceStarter {
        return serviceStarter
    }

}
